src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -35,6 +35,495 @@ extern "C" {
 using namespace gpdxl;
 using namespace gpmd;
 
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::GroupingListContainsPrimaryKey
+//
+//	@doc:
+//		Checks if list grouping_list contains all columns from conkeys of
+//		Primary Key in relation conrelid. grouping_list is treated as flat
+//		list of SortGroupClauses, other nodes are ignored.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
+											   List *grouping_list,
+											   Bitmapset *conkeys, Oid conrelid)
+{
+	BOOL result = false;
+	Bitmapset *grouping_set = NULL;
+	ListCell *lgc;
+
+	// Check if conkeys set is a subset of grouping columns set.
+	ForEach(lgc, grouping_list)
+	{
+		Node *grpcl = (Node *) lfirst(lgc);
+
+		if (NULL == grpcl || !IsA(grpcl, SortGroupClause))
+		{
+			continue;
+		}
+
+		SortGroupClause *sgc = (SortGroupClause *) grpcl;
+		Node *expr = gpdb::GetSortGroupClauseExpr(sgc, query->targetList);
+
+		GPOS_ASSERT(IsA(expr, Var));
+
+		Var *var = (Var *) expr;
+
+		RangeTblEntry *rte =
+			(RangeTblEntry *) gpdb::ListNth(query->rtable, var->varno - 1);
+
+		GPOS_ASSERT(NULL != rte);
+
+		if (gpdb::BmsIsMember(var->varattno, conkeys) &&
+			var->varlevelsup == 0 && rte->relid == conrelid)
+		{
+			grouping_set = gpdb::BmsAddMember(grouping_set, var->varattno);
+		}
+	}
+
+	result = gpdb::BmsIsSubset(conkeys, grouping_set);
```

Comment:
How about such small optimization?
```diff
diff --git a/src/backend/gpopt/translate/CQueryMutators.cpp b/src/backend/gpopt/translate/CQueryMutators.cpp
index 3ae641428e3..2dae0b0781a 100644
--- a/src/backend/gpopt/translate/CQueryMutators.cpp
+++ b/src/backend/gpopt/translate/CQueryMutators.cpp
@@ -50,9 +50,8 @@ CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
 											   List *grouping_list,
 											   Bitmapset *conkeys, Oid conrelid)
 {
-	BOOL result = false;
-	Bitmapset *grouping_set = NULL;
 	ListCell *lgc;
+	int i = 0;
 
 	// Check if conkeys set is a subset of grouping columns set.
 	ForEach(lgc, grouping_list)
@@ -79,15 +78,11 @@ CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
 		if (gpdb::BmsIsMember(var->varattno, conkeys) &&
 			var->varlevelsup == 0 && rte->relid == conrelid)
 		{
-			grouping_set = gpdb::BmsAddMember(grouping_set, var->varattno);
+			i++;
 		}
 	}
 
-	result = gpdb::BmsIsSubset(conkeys, grouping_set);
-
-	gpdb::BmsFree(grouping_set);
-
-	return result;
+	return bms_num_members(conkeys) == i;
 }
 
 //---------------------------------------------------------------------------
```

Reply:
In this case it will not handle column duplicates in grouping sets, for ex:
```
 postgres=# explain select b from test_table1 group by grouping sets((a), ());
                                                      QUERY PLAN                                                       
-----------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice4; segments: 3)  (cost=0.00..1293.00 rows=26 width=8)
   ->  Sequence  (cost=0.00..1293.00 rows=9 width=8)
         ->  Shared Scan (share slice:id 4:0)  (cost=0.00..431.00 rows=9 width=1)
               ->  Materialize  (cost=0.00..431.00 rows=9 width=1)
                     ->  Seq Scan on test_table1  (cost=0.00..431.00 rows=9 width=10)
         ->  Redistribute Motion 1:3  (slice3)  (cost=0.00..862.00 rows=26 width=8)
               ->  Append  (cost=0.00..862.00 rows=9 width=8)
                     ->  Result  (cost=0.00..431.00 rows=1 width=8)
                           ->  Aggregate  (cost=0.00..431.00 rows=1 width=1)
                                 ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.00 rows=1 width=1)
                                       ->  Aggregate  (cost=0.00..431.00 rows=1 width=1)
                                             ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=9 width=1)
                     ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..431.00 rows=25 width=6)
                           ->  Result  (cost=0.00..431.00 rows=9 width=6)
                                 ->  GroupAggregate  (cost=0.00..431.00 rows=9 width=6)
                                       Group Key: share0_ref3.a, share0_ref3.b
                                       ->  Sort  (cost=0.00..431.00 rows=9 width=10)
                                             Sort Key: share0_ref3.a, share0_ref3.b
                                             ->  Shared Scan (share slice:id 2:0)  (cost=0.00..431.00 rows=9 width=10)
 Optimizer: Pivotal Optimizer (GPORCA)
(20 rows)

postgres=# 
postgres=# explain select b from test_table1 group by grouping sets((a,a), ());
INFO:  GPORCA failed to produce a plan, falling back to planner
DETAIL:  Query-to-DXL Translation: No attribute entry found due to incorrect normalization of query
                                                    QUERY PLAN                                                    
------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=5.44..6.42 rows=66 width=48)
   ->  GroupAggregate  (cost=5.44..6.42 rows=22 width=48)
         Group Key: "rollup".prelim_aggref_1, "rollup"."grouping", "rollup"."group_id"
         ->  Subquery Scan on "rollup"  (cost=5.44..5.64 rows=6 width=48)
               ->  Sort  (cost=5.44..5.48 rows=6 width=48)
                     Sort Key: rollup_1.prelim_aggref_1, (Grouping), (group_id())
                     ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=3.83..5.12 rows=6 width=48)
                           Hash Key: rollup_1.prelim_aggref_1, (Grouping), group_id()
                           ->  GroupAggregate  (cost=3.83..4.80 rows=6 width=48)
                                 Group Key: rollup_1."grouping", rollup_1."group_id", rollup_1.prelim_aggref_1
                                 ->  Subquery Scan on rollup_1  (cost=3.83..4.46 rows=9 width=48)
                                       ->  GroupAggregate  (cost=3.83..4.21 rows=9 width=48)
                                             Group Key: test_table1.a
                                             ->  Sort  (cost=3.83..3.89 rows=9 width=10)
                                                   Sort Key: test_table1.a
                                                   ->  Seq Scan on test_table1  (cost=0.00..3.25 rows=9 width=10)
 Optimizer: Postgres query optimizer
(17 rows)
```

but I think we can get rid of ```BmsIsMember``` call, as at the end we are checking for subset relation, not equality.